src/main/java/org/spongepowered/common/mixin/core/network/NetHandlerPlayServerMixin.java

```diff
@@ -396,73 +397,49 @@ private boolean throwMoveEvent(final EntityPlayerMP playerMP, final CPacketPlaye
-            Location<World> fromLocation = player.getLocation();
+            Vector3d fromPosition = player.getPosition();
             if (this.impl$lastMoveLocation != null) {
-                fromLocation = this.impl$lastMoveLocation;
-            }
-
-            Location<World> toLocation = new Location<>(player.getWorld(), packetIn.x, packetIn.y, packetIn.z);
-            Vector3d toRotation = new Vector3d(packetIn.pitch, packetIn.yaw, 0);
-
-            // Minecraft does the same with rotation when it's only a positional update
-            boolean positionOnly = packetIn.moving && !packetIn.rotating;
-            if (positionOnly) {
-                // Correct the new rotation to match the old rotation
-                toRotation = fromRotation;
-
-                positionOnly = !toLocation.getPosition().equals(fromLocation.getPosition()) && ShouldFire.MOVE_ENTITY_EVENT_POSITION;
+                fromPosition = this.impl$lastMoveLocation.getPosition();
             }
+            final Vector3d toPosition = packetIn.moving ? Vector3d.from(packetIn.x, packetIn.y, packetIn.z) : fromPosition;
 
-            // Minecraft sends a 0, 0, 0 position when rotation only update occurs, this needs to be recognized and corrected
-            boolean rotationOnly = !packetIn.moving && packetIn.rotating;
-
-            if (rotationOnly) {
-                // Correct the to location so it's not misrepresented to plugins, only when player rotates without moving
-                // In this case it's only a rotation update, which isn't related to the to location
-                fromLocation = player.getLocation();
-                toLocation = fromLocation;
+            final double deltaSquared = toPosition.distanceSquared(fromPosition);
 
-                rotationOnly = ShouldFire.ROTATE_ENTITY_EVENT;
-            }
 
-            if (packetIn.moving && packetIn.rotating) {
-                positionOnly = !toLocation.getPosition().equals(fromLocation.getPosition()) && ShouldFire.MOVE_ENTITY_EVENT_POSITION;
-                rotationOnly = ShouldFire.ROTATE_ENTITY_EVENT;
-            }
-
-            mixinPlayer.bridge$setVelocityOverride(toLocation.getPosition().sub(fromLocation.getPosition()));
+            final Vector3d fromRotation = player.getRotation();
+            final Vector3d toRotation = packetIn.rotating ? Vector3d.from(packetIn.pitch, packetIn.yaw, 0) : fromRotation;
 
-            final double deltaSquared = toLocation.getPosition().distanceSquared(fromLocation.getPosition());
             final double deltaAngleSquared = fromRotation.distanceSquared(toRotation);
 
+            mixinPlayer.bridge$setVelocityOverride(toPosition.sub(fromPosition));
+
             // These magic numbers are sad but help prevent excessive lag from this event.
             // eventually it would be nice to not have them
-            if (deltaSquared > ((1f / 16) * (1f / 16)) || deltaAngleSquared > (.15f * .15f)) {
```

Comment:
16 -> 16f, this was otherwise returning zero. I'd be tempted to make these numbers named constants.